plutus-core/plutus-ir/src/PlutusIR/Transform/NewLetFloat.hs

```diff
@@ -0,0 +1,478 @@
+              -> Writer (FloatTable tyname name uni fun a) (Binding tyname name uni fun a)
+    goBinding = \case
+        TermBind x s d t -> TermBind x s d <$> go t
+        -- no term inside here, nothing to do
+        b@TypeBind{}     -> pure b
+        b@DatatypeBind{} -> pure b
+
+-- | The 3rd and last pass that, given the result of 'removeLets', places the lets back (floats) at the right marked positions.
+floatBackLets :: forall tyname name uni fun a term re.
+                ( term~Term tyname name uni fun a
+                , re~Reader Depth
+                , PLC.HasUnique tyname PLC.TypeUnique, PLC.HasUnique name PLC.TermUnique)
+              => term -- the cleanedup, reducted term
+              -> FloatTable tyname name uni fun a -- the lets to be floated
+              -> term -- the final, floated, and correctly-scoped term
+floatBackLets term fTable =
+    -- our reader context is only the depth this time.
+    flip runReader topDepth $ goTop term
+  where
+
+    goTop, go :: term -> re term
+
+    -- first check if we have something to float in the top, then start go
+    goTop = floatLam topUnique <=< go
+
+    go = \case
+        -- lam/Lam anchor, increase depth
+        LamAbs a n ty tBody -> local (+1) $
+            LamAbs a n ty <$> (floatLam (n^.PLC.theUnique) =<< go tBody)
+        -- lam/Lam anchor, increase depth
+        TyAbs a n k tBody -> local (+1) $
+            TyAbs a n k <$> (floatLam (n^.PLC.theUnique) =<< go tBody)
+        -- Unfloatable-let anchor, increase depth
+        Let a r bs@(representativeBindingUnique -> letU) tIn -> local (+1) $ do
+            -- note that we do not touch the recursivity of an unfloatable-let
+            Let a r
+                <$> (floatRhs letU =<< traverseOf (traversed.bindingSubterms) go bs)
+                -- act the same as lam/Lam: float right inside
+                <*> (floatLam letU =<< go tIn)
+
+        -- descend
+        t                  -> t & termSubterms go
+
+    floatLam :: PLC.Unique -> term -> re term
+    floatLam lamU =
+        -- make a brand new let-group comprised of all the floatable lets just inside the lam/Lam/letInTerm
+        floatAt (\d -> Pos d lamU LamBody) makeNewLet
+
+    floatRhs :: (binds~ NE.NonEmpty (Binding tyname name uni fun a))
+             => PLC.Unique -> binds -> re binds
+    floatRhs letU =
+        -- we don't know from which rhs the floatable-let(s) came from originally,
+        -- so we instead are going to "squeeze" *AT THE SAME LEVEL* the floatable-let bindings together with the unfloatable let-group's bindings
+        floatAt (\d -> Pos d letU LetRhs) squeezeBindings
+
+    floatAt :: (Depth -> Pos) -- ^ floating position
+            -> (NE.NonEmpty (LetNaked tyname name uni fun a) -> c -> c) -- ^ floating strategy
+            -> c -- ^ term or bindings to float around
+            -> re c -- ^ the combined result
+    floatAt toPos floatFunc termOrBindings = do
+        herePos <- toPos <$> ask
```

Comment:
I think it would be simpler to ask for the position in the calling functions and just get given a position here rather than a function.